internal/server/singleprocess/service_logs.go

```diff
@@ -305,14 +304,160 @@ func (s *service) GetLogStream(
 		)
 	}
 
+	return s.sendInstanceLogs(srv.Context(), log, srv, instanceFunc, req.LimitBacklog)
+}
+
+// Used to reduce the functional surface area of sendInstanceLogs
+type batchSender interface {
+	Send(batch *pb.LogBatch) error
+}
+
+// sendInstanceLogs calls instanceFunc and coordinates sending both the known
+// log entries (ie ones that are already stored on the server) to the waiting
+// client, as well as blocking for all new generate log entries that are created
+// currently known as well as future instances that are returned by instanceFunc.
+func (s *service) sendInstanceLogs(
+	ctx context.Context,
+	log hclog.Logger,
+	sender batchSender,
+	instanceFunc func(ws memdb.WatchSet) ([]*streamRec, error),
+	backlog int32,
+) error {
 	// We keep track of what instances we already have readers for here.
 	var instanceSetLock sync.Mutex
 	instanceSet := make(map[string]struct{})
 
+	// For values returned by LogMerge, we want to be able to map back to
+	// the stream that the reader was for, so we can include the instanceid.
+	// This map lets us do that.
+	readerToInstance := make(map[*logbuffer.Reader]*streamRec)
+
+	// Step 1: we use log merge to read all the known entries from existing
+	// instances. This will never block, it will just weave the log entries
+	// that are already known together and then send them back to the client.
+
+	// Get all current records
+	ws := memdb.NewWatchSet()
+	records, err := instanceFunc(ws)
+	if err != nil {
+		return err
+	}
+	log.Trace("instances loaded", "len", len(records))
+
+	var readers []logbuffer.MergeReader
+
+	for _, record := range records {
+		r := record.LogBuffer.Reader(backlog)
+		readerToInstance[r] = record
+
+		readers = append(readers, r)
+	}
+
+	lm := logbuffer.NewMerger(readers...)
+
+	lines := make([]*pb.LogBatch_Entry, 60)
+
+	// Read out all the log entries from LogMerge. This never blocks waiting
+	// for new log entries, it will simply let each reader output all known
+	// entries and then loop exits.
+	for {
+		entries, err := lm.ReadNext(60)
```

Comment:
This log mergers usage ended up being:
https://ih1.redbubble.net/image.1401903701.4969/flat,750x,075,f-pad,750x1000,f8f8f8.jpg